### ISSUE TITLE
Error on binding when routing

### ISSUE THREAD

**Issue by AStoker** [code-chris/aurelia-knockout#5]
Hello, it's me again :)  
There seems to be an issue with binding multiple times when routing. The setup is having a container which contains a nav bar and a router view. This container is using the `knockout` attribute, since it has knockout in it. However, this container also routes to a page that uses knockout and has a different view model.  
With the way the router activates and binds when routing, the routed view/view model gets bound first (works just fine). But then the container gets bound, and suddenly we get a knockout error saying `ERROR [app-router] Error: You cannot apply bindings multiple times to the same element.` because this container has a div in it that is already bound.  
A possible solution to this is provide some means to stop the binding from taking place inside a `<router-view>` element, limiting the binding to the immediate view/view model.

**Comment by AStoker** [comment]
Additional note, this only happens when _routing_ to a view that has the `knockout` attribute. When the user loads up the page directly that has this kind of setup, the router actually binds from the container down (so we don't have overriding binding contexts). But when routing, the router binds from the last child up to the container (now overriding binding contexts).

**Comment by AStoker** [comment]
A current work around is based on this post: http://www.knockmeout.net/2012/05/quick-tip-skip-binding.html
If you apply the `stopBindings` binding to the `<router-view>`, it will stop the issue of the overriding contexts. So the question would be if can can identify the elements that Aurelia automatically binds new contexts to, and not bind those elements with knockout.

**Comment by ckotzbauer** [comment]
Only to understand the structure correctly: The view which owns the `<router-view>` element has knockout bindings. And the error occurs if you route to a view which has also such bindings:

**app.html**

``` html
<div knockout>
    <h1 data-bind="text: header"></h1>
    <button data-bind="click: switchRoute">Switch</button>
    <router-view></router-view>
</div>
```

With this scenario I can't reproduce the issue, because the app.html is not bound again.

**Comment by AStoker** [comment]
If your **app.html** looked like that, and then you have a route that's navigated to, and that file (let's call it **child.html**) has the same knockout binding attribute, that's when it fails. There are two different view models that need to be bound. The app.html needs to be bound to the app.js, and the child.html needs to be bound to the child.js.
Are you saying you can't reproduce that setup?

**Comment by AStoker** [comment]
I can try and get a skeleton repo up if I need to.  
Also, by the way, love what you're doing here. It's saving my backside in this process of converting a Durandal page to Aurelia :)

**Comment by ckotzbauer** [comment]
Yes, that's exactly the setup I tried. I can switch between my three child models as often as I would like without any errror.

It would be nice if you can create a small skeleton repo.

**Comment by AStoker** [comment]
Okay, more details. I tried it on the skeleton project with Aurelia, and it worked fine... For the first layer of navigation.  
If you put the `knockout` binding on the app.html, as well as the child-router.html, it works fine if you navigate to the child router. But if you navigate to the child routers child route (the inception route if you will ;)), well, it behaves correctly. But now, refresh the page with the nested child routes (url looks something like this `http://localhost:9000/#/child-router/child-router`) and you'll get the binding error.

**Comment by AStoker** [comment]
Repro Repo: https://github.com/AStoker/aurelia-knockout-issues

**Comment by ckotzbauer** [comment]
Thanks for the skeleton. I can reproduce it.
I've found a possible solution: I can simply unbind the the node before apply the bindings again. Even, if this is not really clean because it unbinds the node and all childs (including them which were bound recently). But it seems to fix this issue.

**Comment by AStoker** [comment]
Would that remove any bindings specific to the children (making the children loose their initial binding view-model and now be bound to the parent)?

**Comment by ckotzbauer** [comment]
Yes, this removes all knockout bindings (no aurelia bindings) from any child element in the DOM. I tried it and it seems to be, that the child views are bound to the parent until the knockout custom attribute of the child view is processed again which does not happen automatically...

**Comment by AStoker** [comment]
What happens if that child binding is relying on something that doesn't exist on the parent, and causes the binding to fail? It does seem a bit iffy to bind the whole thing to a single VM instead of the individual parts.  
Would it not be possible to determine which elements have their own binding instance and not have knockout bind those children?

**Comment by ckotzbauer** [comment]
If pushed some changes to fix this behavior without unbinding hard. The solution was to skip bindings in DOM nodes which are already bound like described in the blog post you've mentioned above. This is released as 1.0.1-beta.1 without npm package. Please give feedback!

**Comment by AStoker** [comment]
Works like a charm!

**Comment by ckotzbauer** [comment]
Released in 1.0.1
